### PR TITLE
Fix issue when isolating libraries when file not ready

### PR DESCRIPTION
**Pull request by Hazhzeng**
## Description

Fixes an issue in Python 3.9 Linux Consumption plan. This occurs when the customer's `azure.functions` library is not presented as we tried to clear out its cache prematurely. This will result in a longer cold start time since the function host will restart several times to wait for the customer's content before doing specialization. The worst scenario will cause the website goes into a 404 state for a few seconds during the code start.

```
Result:
Failure Exception:
KeyError: 'azure.functions'
Stack:
   File '/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/dispatcher.py', line 482, in _handle__function_environment_reload_request     DependencyManager.reload_azure_google_namespace(
   File '/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/utils/dependency.py', line 155, in reload_azure_google_namespace     cls.reload_all_namespaces_from_customer_deps(cx_working_dir)
   File '/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/utils/dependency.py', line 207, in reload_all_namespaces_from_customer_deps     cls._remove_from_sys_path(cls.worker_deps_path)
   File '/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/utils/dependency.py', line 259, in _remove_from_sys_path     cls._clear_path_importer_cache_and_modules(path)
   File '/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/utils/dependency.py', line 278, in _clear_path_importer_cache_and_modules     cls._remove_module_cache(path)
   File '/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/utils/dependency.py', line 353, in _remove_module_cache     elif getattr(module, '__path__', None) and getattr(
   File '<frozen importlib._bootstrap_external>', line 1255, in __len__
   File '<frozen importlib._bootstrap_external>', line 1234, in _recalculate
   File '<frozen importlib._bootstrap_external>', line 1230, in _get_parent_path 
```

### Discussion
1. When customer's trigger name collided with customer's libraries, should give a warning. 

### Mitigation
This is currently impacting the customers using Python 3.9 Linux Consumption. Setting the `PYTHON_ISOLATE_WORKER_DEPENDENCIES` to `0` will mitigate this issue for now.

### Validation:
1. Use `pytest  --instafail --cov=./azure_functions_worker --cov-report xml --cov-branch .\tests\unittests\test_extension.py .\tests\unittests\test_utilities_dependency.py --randomly-seed=3973973583` to introduce unexpected module loading order. This should fails with `azure.functions key not found` before this fix.

### Also found that improper namespacing in worker nuget artifact:
The \_\_init\_\_.py file should only be placed in https://github.com/Azure/azure-functions-python-library source code. However, this shouldn't be included in a release package.
image

According to Python namespace specification https://www.python.org/dev/peps/pep-0420/#specification. Quote:
> Namespace packages cannot contain an __init__.py. As a consequence, pkgutil.extend_path and pkg_resources.declare_namespace become obsolete for purposes of namespace package creation. There will be no marker file or directory for specifying a namespace package.

---

### PR information

- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] If applicable, the PR references the bug/issue that it fixes in the description.
- [x] New Unit tests were added for the changes made and CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-functions-python-worker/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Hazhzeng** [comment]
Validation:

Without the fix from https://github.com/Azure/azure-functions-python-library/pull/90:
image

With the fix from https://github.com/Azure/azure-functions-python-library/pull/90:
image


Tested with the following mesh docker image and command:
[run_host_mesh_custom.ps1](https://github.com/Azure/azure-functions-python-worker/files/6590534/run_host_mesh_custom.txt)
```powershell
.\run_host_mesh_custom.ps1 -ImageName mcr.microsoft.com/azure-functions/mesh:3.0.15828-python3.9 -ScriptRoot /C/Users/hazeng/Projects/Functions/PythonNoReq -PythonVersion 3.9
```

**Comment by Hazhzeng** [comment]
Validation on Linux Consumption (image: mcr.microsoft.com/azure-functions/mesh:3.0.15828-python3.9)
Without the change, able to reproduce the issue:
image
image

With the latest Worker and Library mounted, able to mitigate this issue. Confirmed with multiple runs:
image
image

**Comment by Hazhzeng** [comment]
Validation on Linux Consumption Python 3.8 (image: mcr.microsoft.com/azure-functions/mesh:3.0.15828-python3.8)
Without any changes, feature flag disabled by default (expect to succeed):
image
image

With the latest worker and library mounted, feature flag disabled by default (expect to succeed):
image
image

**Comment by Hazhzeng** [comment]
Validation on Linux Dedicated Python 3.8 (image: mcr.microsoft.com/azure-functions/python:3.0.15828-python3.8)
With the original image, feature flag is disabled by default. Customer brings `grpcio-1.35.0` / Worker has `grpcio-1.33.2` (collision detected, function is not able to start):
image

With the original image, setting the PYTHON_ISOLATE_WORKER_DEPENDENCIES = 1, function app is now working:
image
image

With the latest worker update, feature flag is disabled by default. (collision detected, function is not able to start):
image

With the latest worker update, enable the feature flag, function is recovered:
image
image

**Comment by Hazhzeng** [comment]
A new round of test with the \_\_init\_\_.py removed from worker dependencies azure/ folder.
Python 3.9 consumption:
image

Python 3.9 dedicated:
image

Python 3.8 consumption:
image

Python 3.8 dedicated (expect to fail, since collision in [customer grpcio-1.35.0 / worker grpcio-1.33.2] library):
image

File: azure_functions_worker/constants.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 # Platform Environment Variables
 AZURE_WEBJOBS_SCRIPT_ROOT = "AzureWebJobsScriptRoot"
+CONTAINER_NAME = "CONTAINER_NAME"
```
Comment (vrdmr) [inline azure_functions_worker/constants.py:21]: Can we have this constant in dependency.py itself? This isn't being used globally.

File: azure_functions_worker/utils/dependency.py [hunk 2]
```diff
@@ -112,24 +123,50 @@ def prioritize_customer_dependencies(cls):
...
+        2. worker_deps_path
+        3. cx_working_dir
         """
+        # Try to get the latest customer's working directory
+        # cx_working_dir => cls.cx_working_dir => AzureWebJobsScriptRoot
+        working_directory: str = ''
+        if cx_working_dir:
+            working_directory: str = os.path.abspath(cx_working_dir)
+        if not working_directory:
+            working_directory = cls.cx_working_dir
```
Comment (vrdmr) [inline azure_functions_worker/utils/dependency.py:137]: Can there ever be a case that cx_working_dir isn't a property in cls?

File: azure_functions_worker/utils/dependency.py [hunk 3]
```diff
@@ -112,24 +123,50 @@ def prioritize_customer_dependencies(cls):
...
         # Otherwise, it will break some really old function apps, those
         # don't have azure-functions module in .python_packages
         # https://github.com/Azure/azure-functions-core-tools/pull/1498
         cls._add_to_sys_path(cls.worker_deps_path, False)
 
-        logger.info(f'Start using customer dependencies {cls.cx_deps_path}')
+        # The modules defined in customer's working directory should have the
+        # least priority since we uses the new folder structure.
+        # Please check the "Message to customer" section in the following PR:
+        # https://github.com/Azure/azure-functions-python-worker/pull/726
```
Comment (vrdmr) [inline azure_functions_worker/utils/dependency.py:163]: Can we make a pinned issue as well (or check if this is added to Python Dev Docs itself)?

File: azure_functions_worker/utils/dependency.py [hunk 4]
```diff
@@ -112,24 +123,50 @@ def prioritize_customer_dependencies(cls):
...
+        cls._add_to_sys_path(working_directory, False)
+
+        logger.info('Finished prioritize_customer_dependencies')
 
     @classmethod
-    def reload_azure_google_namespace(cls, cx_working_dir: str):
+    def reload_customer_libraries(cls, cx_working_dir: str):
         """Reload azure and google namespace, this including any modules in
```
Comment (vrdmr) [inline azure_functions_worker/utils/dependency.py:170]: This is not only doing Azure and Google right? Comment should reflect this now.

File: pack/templates/nix_env_gen.yml [hunk 5]
```diff
@@ -12,7 +12,10 @@ steps:
   inputs:
     disableAutoCwd: true
     scriptPath: 'pack/scripts/nix_deps.sh'
+# The following task is disabled since Python 3.9 to fix azure/ namespace
+# https://github.com/Azure/azure-functions-python-worker/pull/854
```
Comment (vrdmr) [inline pack/templates/nix_env_gen.yml:16]: Will we re-enable this particular task later?

File: tests/endtoend/dependency_functions/report_dependencies/__init__.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import azure.functions as func
+import google.protobuf as proto
+import grpc
+
+# Load dependency manager from customer' context
+from azure_functions_worker.utils.dependency import DependencyManager as dm
+
+
+def main(req: func.HttpRequest) -> func.HttpResponse:
+    """This function is an HttpTrigger to check if the modules are loaded from
```
Comment (vrdmr) [inline tests/endtoend/dependency_functions/report_dependencies/__init__.py:13]: Nice test! :)

**Review by vrdmr — APPROVED** [review]
minor comments.

LGTM

File: pack/templates/nix_env_gen.yml (hunk 5 above)
Comment (Hazhzeng) [reply pack/templates/nix_env_gen.yml:16]: So this is an actual fix for Python 3.9 and afterward.

**Comment by vivek-bsn** [comment]
@Hazhzeng Currently the mitigation provided works to avoid 404 errors. Any update on when this fix is expected to roll out to the production so that the app setting 'PYTHON_ISOLATE_WORKER_DEPENDENCIES ' can be removed?

**Comment by vrdmr** [comment]
The fix is currently being rolled out and is halfway across all our fleet. It should be completed by mid-next week to all the geo-locations.

**Comment by vivek-bsn** [comment]
Once the fix is deployed, we can remove the setting 'PYTHON_ISOLATE_WORKER_DEPENDENCIES' to 0 right? (or) do we need to retain it in the app settings?

**Pull request merged**